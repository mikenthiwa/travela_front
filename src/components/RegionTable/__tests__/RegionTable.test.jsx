import React from 'react';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { RegionTable } from '../index';

const props = {
  regions: [
    {
      id: 1002,
      region: 'West Africa',
      'description': 'Nigeria',
    },
    {
      id: 1001,
      region: 'East Africa',
      description: 'Kenya, Uganda and Rwanda',
    }
  ],
  pagination: {
    pageCount: 3,
    currentPage: 3,
    totalCount: 21
  },
  deleteRegion: jest.fn(),
  handleEditRegion: jest.fn(),
  deleteTravelRegion: jest.fn(),

};
const state = {
  currentPage: 1,
  searchTerm: '',
  pageStart: 0,
  showDeleteModal: false,
  isDeleting: false,
  regionId: '',
};
let wrapper;
beforeEach(() => {
  wrapper = mount(<MemoryRouter><RegionTable {...props} /></MemoryRouter>);
  jest.clearAllMocks();
});

it('renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
describe('RegionTable ', () => {
  const wrapper = shallow(<RegionTable {...props} />);
  it('should render the table with regions', () => {
    expect(wrapper.find('table.mdl-data-table').length).toBe(1);
    expect(wrapper.find('.table__row').length).toBe(2);
  });

  it('should render a div when there are regions', () => {
    wrapper.setProps({ regions: [] });
    expect(wrapper.find('table.mdl-data-table').length).toBe(0);
    expect(wrapper.find('div.table__requests--empty').length).toBe(1);
  });
});

describe('HandleSearch Region', () => {
  it('should search for regionTable', () => {
    const wrapper = mount(<MemoryRouter><RegionTable {...props} /></MemoryRouter>);
    const search = wrapper.find('.input-field');
    search.simulate('change', { target: { value: 'ug' } });
    expect(wrapper.find('.mdl-data-table')).toHaveLength(1);
  });

  it('should visit the next page', () => {
    const wrapper = mount(<MemoryRouter><RegionTable {...props} /></MemoryRouter>);
    const nextBtn = wrapper.find('.pagination__button').last();
    nextBtn.simulate('click');
    expect(wrapper.find('.mdl-data-table')).toHaveLength(1);
  });

  it('should visit the previous page', () => {
    const wrapper = mount(<MemoryRouter><RegionTable {...props} /></MemoryRouter>);
    const nextBtn = wrapper.find('.pagination__button').first();
    nextBtn.simulate('click');
    expect(wrapper.find('tbody')).toHaveLength(1);
  });
});

describe('Edit Travel Region', () => {
  it('should open the edit travel regions modal', () => {
    const wrapper = mount(
      <MemoryRouter>
        <RegionTable {...props} />
      </MemoryRouter>
    );
    wrapper.setProps({
      shouldOpen: true,
      modalType: 'edit travel regions'
    });

    wrapper.find('.edit-region .edit').at(0).simulate('click');

    expect(wrapper.find('Modal').first().props().visibility).toEqual('invisible');
  });
  it('should call toggleDeleteModal', () => {
    const shallowWrapper = shallow(<RegionTable />);
    const defaultState = {
      currentPage: 1,
      searchTerm: '',
      pageStart: 0,
      showDeleteModal: false,
      isDeleting: false,
      regionId: '',
    };
    shallowWrapper.instance().setState(defaultState);
    const event = {
      preventDefault: jest.fn(),
    };
    sinon.spy(shallowWrapper.instance(), 'toggleDeleteModal');
    sinon.spy(shallowWrapper.instance(), 'closeModal');
    sinon.spy(shallowWrapper.instance(), 'onPageChange');
    shallowWrapper.instance().closeModal();
    shallowWrapper.instance().toggleDeleteModal(event);
    shallowWrapper.instance().onPageChange(event);
    expect(shallowWrapper.instance().toggleDeleteModal.calledOnce).toEqual(true);
    expect(shallowWrapper.instance().closeModal.calledOnce).toEqual(true);
    expect(shallowWrapper.instance().onPageChange.calledOnce).toEqual(true);
  });

  it('should not call toggleDeleteModal', () => {
    const shallowWrapper = shallow(<RegionTable />);
    const defaultState = {
      showDeleteModal: true
    };
    shallowWrapper.instance().setState(defaultState);
    const event = {
      preventDefault: jest.fn(),
    };
    sinon.spy(shallowWrapper.instance(), 'toggleDeleteModal');
    shallowWrapper.instance().toggleDeleteModal(event);
    expect(shallowWrapper.contains('showDeleteModal')).toEqual(false);
  });

  it('should call deleteTravelRegion', () => {
    const id = { target: { regionId: 1002 } };
    const wrapper = mount(<MemoryRouter><RegionTable {...props} /></MemoryRouter>);
    const instance = wrapper.find(RegionTable).instance();
    instance.setState({ regionId: 1002 });
    expect(instance.deleteTravelRegion(id)).toBeCalled;
  });
});
