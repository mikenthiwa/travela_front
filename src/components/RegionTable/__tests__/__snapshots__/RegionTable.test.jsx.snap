// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="table__container user-roles"
>
  <div
    className="search-container"
  >
    <span
      className="icon"
    >
      Region
    </span>
    <input
      className="input-field"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      type="button"
    >
      <span
        className="fa fa-search search-icon"
      />
    </button>
  </div>
  <table
    className="mdl-data-table mdl-js-data-table table__requests"
  >
    <thead>
      <tr>
        <th
          className="mdl-data-table__cell--non-numeric bb-md-0 table__head freeze request_id"
        >
          Region Name
        </th>
        <th
          className="mdl-data-table__cell--non-numeric table__head pl-sm-100d description-left"
        >
          Description
        </th>
      </tr>
    </thead>
    <tbody
      className="table__body approvals_table_body"
    >
      <tr
        className="table__row table__effects"
        key="1002"
      >
        <td
          className="mdl-data-table__cell--non-numeric table__data freeze table__data-pointer"
        >
          <a
            className="table__data--link button-outline table__id-pointer"
            href="/settings/travel-region/1002"
            onClick={[Function]}
          >
            West Africa
          </a>
        </td>
        <td
          className="mdl-data-table__cell--non-numeric table__data pl-sm-120"
        >
          Nigeria
        </td>
        <td
          className="table__data edit-region"
        >
          <span>
            <i
              className="fa fa-ellipsis-v on"
              onClick={[Function]}
              role="presentation"
            />
            <div
              className="context-menu table__menu-container false"
            >
              <ul
                className="table__menu-list"
              >
                <li
                  className="edit"
                  onClick={[Function]}
                  role="presentation"
                >
                  <span>
                    Edit
                  </span>
                </li>
                <li
                  className="delete"
                  onClick={[Function]}
                  role="presentation"
                >
                  <span>
                    Delete
                  </span>
                </li>
              </ul>
            </div>
          </span>
        </td>
      </tr>
      <tr
        className="table__row table__effects"
        key="1001"
      >
        <td
          className="mdl-data-table__cell--non-numeric table__data freeze table__data-pointer"
        >
          <a
            className="table__data--link button-outline table__id-pointer"
            href="/settings/travel-region/1001"
            onClick={[Function]}
          >
            East Africa
          </a>
        </td>
        <td
          className="mdl-data-table__cell--non-numeric table__data pl-sm-120"
        >
          Kenya, Uganda and Rwanda
        </td>
        <td
          className="table__data edit-region"
        >
          <span>
            <i
              className="fa fa-ellipsis-v on"
              onClick={[Function]}
              role="presentation"
            />
            <div
              className="context-menu table__menu-container false"
            >
              <ul
                className="table__menu-list"
              >
                <li
                  className="edit"
                  onClick={[Function]}
                  role="presentation"
                >
                  <span>
                    Edit
                  </span>
                </li>
                <li
                  className="delete"
                  onClick={[Function]}
                  role="presentation"
                >
                  <span>
                    Delete
                  </span>
                </li>
              </ul>
            </div>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <div
      className="pagination"
    >
      <button
        className="pagination__button"
        disabled={true}
        id="previous-button"
        onClick={[Function]}
        type="button"
      >
        <span
          className=""
        >
          Previous
        </span>
        <span
          className=" mdl-cell--hide-desktop mdl-cell--hide-tablet"
        >
          Previous
        </span>
      </button>
      <div
        className="pagination__items"
      >
        <span
          className="pagination__page"
        >
          Page
        </span>
        <div
          className="pagination__current-page"
          id="current-page"
        >
          1
        </div>
        of
        <span
          className="pagination__all-pages"
          id="page-count"
        >
          1
        </span>
      </div>
      <button
        className="pagination__button"
        disabled={true}
        id="next-button"
        onClick={[Function]}
        type="button"
      >
        <span
          className=""
        >
          Next
        </span>
        <span
          className=" mdl-cell--hide-desktop mdl-cell--hide-tablet"
        >
          Next
        </span>
      </button>
    </div>
  </div>
</div>
`;
